cpp/src/arrow/compute/registry.h

```diff
@@ -45,20 +45,42 @@ class FunctionOptionsType;
 /// lower-level function execution.
 class ARROW_EXPORT FunctionRegistry {
  public:
-  ~FunctionRegistry();
+  virtual ~FunctionRegistry();
 
   /// \brief Construct a new registry. Most users only need to use the global
   /// registry
   static std::unique_ptr<FunctionRegistry> Make();
 
+  /// \brief Construct a new nested registry with the given parent. Most users only need
+  /// to use the global registry
+  static std::unique_ptr<FunctionRegistry> Make(FunctionRegistry* parent);
+
+  /// \brief Construct a new nested registry with the given parent. Most users only need
+  /// to use the global registry
+  static std::unique_ptr<FunctionRegistry> Make(std::unique_ptr<FunctionRegistry> parent);
```

Comment:
A function that accepts a unique_ptr should take ownership of that unique pointer.  That does not seem to be what we are doing here.  In this case I think we should probably get rid of this overload.

There is a "maybe owned" pattern we use in a few places where we have a constructor that takes a raw pointer and one that takes a shared_ptr but in that case we have two member variables (e.g. https://github.com/apache/arrow/blob/master/cpp/src/arrow/io/caching.cc#L147)